### PR TITLE
Fix instance binding for empty transformed masks

### PR DESCRIPTION
**Pull request by ternaus**
## Summary

- Remove a bound instance when its transformed mask contains no non-zero pixels, even if a clipped bbox fragment still passes bbox filtering.
- Apply that survival decision to the instance bbox, bbox labels, and keypoints for both `mask` and `masks` layouts.
- Document the behavior for `Compose.instance_binding` and `CopyAndPaste(min_visibility_after_paste=0.0)`.

## Reproduction

An `Affine` transform moves an edge object completely outside the image. Its mask becomes empty, but its transformed axis-aligned bbox leaves a clipped fragment inside the image. Before this change, the bbox fragment kept the entire bound instance alive with an empty mask.

The regression test also includes a second object that stays in frame, so it verifies that only the empty-mask instance is removed and the surviving instance keeps its mask and label.

## Validation

- `100 passed` in `tests/test_instance_binding.py`
- `12,936 passed, 42 skipped, 38 deselected, 9 xfailed, 3 xpassed` in the full non-slow suite
- `tools.quality_gate fast` passed
- `pre-commit run --all-files` passed in the project environment

## Performance

The affected `Affine` pipeline was benchmarked before and after over 256, 512, and 1024 pixel images; 1, 3, and 5 image channels; and 1 or 8 bound masks. The largest observed positive change was 3.2%, below the 5% guardrail.

A separate `NoOp` stress benchmark isolates the required final mask scan. It added about 0.026 ms for one 1024x1024 mask and 0.223 ms for eight 1024x1024 masks. This is the expected cost of checking every bound mask for non-zero pixels when the pipeline itself does otherwise negligible work.

## Summary by Sourcery

Ensure bound instances are removed when their transformed masks become empty so that masks, bboxes, labels, and keypoints stay consistent, and document this behavior in instance binding and copy-paste APIs.

Bug Fixes:
- Remove instances whose transformed masks contain no non-zero pixels even if their clipped bboxes still pass bbox filtering, keeping masks and annotations aligned.
- Update copy-paste behavior so fully occluded primaries with empty masks are dropped even when the visibility threshold is zero.

Enhancements:
- Describe instance_binding behavior in Compose, including how bound masks, bboxes, and keypoints are jointly filtered based on bbox and mask visibility.
- Clarify CopyAndPaste(min_visibility_after_paste) documentation to cover interaction with instance binding and empty masks.

Tests:
- Add regression tests covering instance removal when masks become empty after affine transforms for both `mask` and `masks` layouts.
- Adjust copy-paste tests to assert that fully occluded primaries with empty masks are removed when instance binding is enabled.

### CODE REVIEW

**Comment by ternaus** [comment]
## Follow-up validation and complete benchmark evidence

Commit `0514b80` fixes terminal `ToTensorV2` handling for both bound mask layouts, including `transpose_mask=True`, final bbox filtering, ambiguous `N == H == W` shapes, and all-empty Tensor selections.

### Method

- Apple Silicon macOS, Python 3.13, NumPy 2.5, OpenCV threads fixed to 1.
- CPU process time, GC disabled, warmup before measurement, paired before/after blocks with alternating order.
- Direct table compares the PR's previous NumPy-only reduction with the final NumPy/Tensor-aware helper.
- Compose table compares the same Affine + instance-binding pipeline with the required empty-mask scan disabled versus enabled.
- The one initial noisy Compose cell above 5% (`1024²×3`, 8 masks, `masks`) was rerun for 15 paired blocks × 100 iterations; its median was `5.8492 → 6.0437 ms (+3.32%)`.
- Direct range: **-2.90%…+4.31%**. Compose range: **-0.76%…+4.47%**. No median cell exceeds the 5% guardrail.

### Direct mask-survival helper (ms/call)

| Size | Image ch. | Masks | Layout | Before | After | Change |
|---:|---:|---:|:---|---:|---:|---:|
| 256² | 1 | 1 | `mask` | 0.00230 | 0.00238 | +3.47% |
| 256² | 1 | 1 | `masks` | 0.00229 | 0.00233 | +1.83% |
| 256² | 1 | 8 | `mask` | 0.38758 | 0.38901 | +0.37% |
| 256² | 1 | 8 | `masks` | 0.01300 | 0.01306 | +0.47% |
| 256² | 3 | 1 | `mask` | 0.00234 | 0.00243 | +3.47% |
| 256² | 3 | 1 | `masks` | 0.00234 | 0.00233 | -0.13% |
| 256² | 3 | 8 | `mask` | 0.41294 | 0.41048 | -0.59% |
| 256² | 3 | 8 | `masks` | 0.01315 | 0.01306 | -0.68% |
| 256² | 5 | 1 | `mask` | 0.00228 | 0.00238 | +4.31% |
| 256² | 5 | 1 | `masks` | 0.00232 | 0.00236 | +1.60% |
| 256² | 5 | 8 | `mask` | 0.38868 | 0.39164 | +0.76% |
| 256² | 5 | 8 | `masks` | 0.01369 | 0.01363 | -0.42% |
| 512² | 1 | 1 | `mask` | 0.00742 | 0.00761 | +2.65% |
| 512² | 1 | 1 | `masks` | 0.00708 | 0.00723 | +2.12% |
| 512² | 1 | 8 | `mask` | 1.53483 | 1.53227 | -0.17% |
| 512² | 1 | 8 | `masks` | 0.05099 | 0.05065 | -0.66% |
| 512² | 3 | 1 | `mask` | 0.00702 | 0.00711 | +1.26% |
| 512² | 3 | 1 | `masks` | 0.00699 | 0.00703 | +0.46% |
| 512² | 3 | 8 | `mask` | 1.55016 | 1.54765 | -0.16% |
| 512² | 3 | 8 | `masks` | 0.05040 | 0.05063 | +0.47% |
| 512² | 5 | 1 | `mask` | 0.00703 | 0.00711 | +1.12% |
| 512² | 5 | 1 | `masks` | 0.00708 | 0.00703 | -0.63% |
| 512² | 5 | 8 | `mask` | 1.55158 | 1.55824 | +0.43% |
| 512² | 5 | 8 | `masks` | 0.04808 | 0.04825 | +0.37% |
| 1024² | 1 | 1 | `mask` | 0.02465 | 0.02424 | -1.67% |
| 1024² | 1 | 1 | `masks` | 0.02479 | 0.02504 | +1.00% |
| 1024² | 1 | 8 | `mask` | 6.33538 | 6.24081 | -1.49% |
| 1024² | 1 | 8 | `masks` | 0.18840 | 0.18844 | +0.02% |
| 1024² | 3 | 1 | `mask` | 0.02439 | 0.02436 | -0.13% |
| 1024² | 3 | 1 | `masks` | 0.02570 | 0.02544 | -0.99% |
| 1024² | 3 | 8 | `mask` | 6.29215 | 6.10983 | -2.90% |
| 1024² | 3 | 8 | `masks` | 0.18856 | 0.18745 | -0.59% |
| 1024² | 5 | 1 | `mask` | 0.02585 | 0.02592 | +0.25% |
| 1024² | 5 | 1 | `masks` | 0.02422 | 0.02425 | +0.13% |
| 1024² | 5 | 8 | `mask` | 6.30142 | 6.31723 | +0.25% |
| 1024² | 5 | 8 | `masks` | 0.19053 | 0.19098 | +0.24% |

### Full Compose/Affine route (ms/call)

| Size | Image ch. | Masks | Layout | Before | After | Change |
|---:|---:|---:|:---|---:|---:|---:|
| 256² | 1 | 1 | `mask` | 0.31179 | 0.32006 | +2.65% |
| 256² | 1 | 1 | `masks` | 0.33286 | 0.33787 | +1.50% |
| 256² | 1 | 8 | `mask` | 0.75183 | 0.76054 | +1.16% |
| 256² | 1 | 8 | `masks` | 0.53887 | 0.55642 | +3.26% |
| 256² | 3 | 1 | `mask` | 0.37589 | 0.37534 | -0.15% |
| 256² | 3 | 1 | `masks` | 0.38217 | 0.38735 | +1.36% |
| 256² | 3 | 8 | `mask` | 0.77740 | 0.79476 | +2.23% |
| 256² | 3 | 8 | `masks` | 0.60692 | 0.61400 | +1.17% |
| 256² | 5 | 1 | `mask` | 0.70636 | 0.71618 | +1.39% |
| 256² | 5 | 1 | `masks` | 0.69486 | 0.70324 | +1.21% |
| 256² | 5 | 8 | `mask` | 1.12833 | 1.13187 | +0.31% |
| 256² | 5 | 8 | `masks` | 0.95661 | 0.96912 | +1.31% |
| 512² | 1 | 1 | `mask` | 0.55989 | 0.57027 | +1.85% |
| 512² | 1 | 1 | `masks` | 0.57143 | 0.57741 | +1.05% |
| 512² | 1 | 8 | `mask` | 2.12631 | 2.17650 | +2.36% |
| 512² | 1 | 8 | `masks` | 1.37125 | 1.42370 | +3.82% |
| 512² | 3 | 1 | `mask` | 0.79747 | 0.81127 | +1.73% |
| 512² | 3 | 1 | `masks` | 0.84470 | 0.85497 | +1.22% |
| 512² | 3 | 8 | `mask` | 2.41845 | 2.47490 | +2.33% |
| 512² | 3 | 8 | `masks` | 1.60468 | 1.67635 | +4.47% |
| 512² | 5 | 1 | `mask` | 2.13527 | 2.13857 | +0.15% |
| 512² | 5 | 1 | `masks` | 2.13170 | 2.14416 | +0.58% |
| 512² | 5 | 8 | `mask` | 3.73613 | 3.78778 | +1.38% |
| 512² | 5 | 8 | `masks` | 2.94576 | 2.98063 | +1.18% |
| 1024² | 1 | 1 | `mask` | 1.56814 | 1.59311 | +1.59% |
| 1024² | 1 | 1 | `masks` | 1.58133 | 1.59855 | +1.09% |
| 1024² | 1 | 8 | `mask` | 7.63226 | 7.88319 | +3.29% |
| 1024² | 1 | 8 | `masks` | 4.78663 | 4.98913 | +4.23% |
| 1024² | 3 | 1 | `mask` | 2.59932 | 2.68838 | +3.43% |
| 1024² | 3 | 1 | `masks` | 2.64165 | 2.68103 | +1.49% |
| 1024² | 3 | 8 | `mask` | 8.94195 | 9.12577 | +2.06% |
| 1024² | 3 | 8 | `masks` | 5.84920 | 6.04370 | +3.32% |
| 1024² | 5 | 1 | `mask` | 8.67009 | 8.60424 | -0.76% |
| 1024² | 5 | 1 | `masks` | 7.95130 | 8.21718 | +3.34% |
| 1024² | 5 | 8 | `mask` | 14.28469 | 14.57744 | +2.05% |
| 1024² | 5 | 8 | `masks` | 11.81143 | 11.97557 | +1.39% |

### Tensor reduction backend candidate (8 masks, ms/call)

The Tensor path uses a zero-copy CPU NumPy view because it beat `torch.any` consistently with both libraries restricted to one thread.

| Size | Layout | `torch.any` | `np.any(tensor.numpy())` | Chosen path improvement |
|---:|:---|---:|---:|---:|
| 256² | `mask` | 0.5733 | 0.3941 | 31.3% |
| 256² | `masks` | 0.0140 | 0.0129 | 7.9% |
| 512² | `mask` | 2.1959 | 1.5564 | 29.1% |
| 512² | `masks` | 0.0556 | 0.0530 | 4.7% |
| 1024² | `mask` | 8.8044 | 6.2050 | 29.5% |
| 1024² | `masks` | 0.2059 | 0.1913 | 7.1% |

### Performance audit record

- Kept the single required full-array mask read; no extra H×W pass was added.
- Preserved the original NumPy fast path, avoiding layout bookkeeping on existing arrays.
- Tensor conversion is zero-copy on the supported CPU route; only the O(N) surviving-index vector is materialized for safe `index_select`.
- Pixel loops, grouped reductions/`bincount`, LUTs, RNG paths, and in-place mutation are not applicable to this boolean per-instance reduction.
- The operation remains in Compose because it encodes instance-binding policy; it is not an Albucore image-processing primitive.

### Validation

- `uv run pytest -m "not slow"`: **12,787 passed, 42 skipped, 38 deselected, 9 xfailed, 3 xpassed**
- `uv run python -m tools.quality_gate fast`: passed, including ruff, mypy, Pyrefly, 1,232 contract tests, CI matrix/shard, benchmark budget, legal integrity, and regression vectors
- `pre-commit run --all-files`: passed

File: albumentations/core/composition.py [hunk 1]
```diff
@@ -748,6 +751,59 @@ def _mirror_keep_mask(
...
+        binding: frozenset[str],
+        instance_count: int,
+    ) -> np.ndarray | None:
+        """Return one boolean per bound instance indicating whether its transformed mask contains at least one non-zero
+        pixel after augmentation.
+        """
+        if "masks" in binding:
+            masks = data.get("masks")
+            if isinstance(masks, np.ndarray) and len(masks) == instance_count:
+                return np.any(masks, axis=tuple(range(1, masks.ndim)))
```
[the remark this answers is not in the document]
Comment (ternaus) [reply albumentations/core/composition.py:809]: Fixed in `0514b80`. Instance binding now recognizes CPU Tensor masks, preserves the correct instance axis for packed and stacked layouts (including `transpose_mask=True`), mirrors final bbox filtering onto Tensor masks, and uses `index_select` for safe empty/boolean selections. Added public Compose regressions for both layouts, both transpose modes, final bbox filtering, ambiguous `N == H == W`, and all-empty outputs. The full non-slow suite and fast quality gate pass.

[the remark this answers is not in the document]
Comment (ternaus) [reply albumentations/core/composition.py:809]: Published every direct-helper and Compose/Affine cell in the [benchmark evidence](https://github.com/albumentations-team/AlbumentationsX/pull/384#issuecomment-5157982839). The direct range is `-2.90%…+4.31%`; the Compose range is `-0.76%…+4.47%`; no median cell exceeds the 5% guardrail. The report also includes the Tensor backend comparison, methodology, performance audit, and exact validation results.

**Comment by ternaus** [comment]
## CPU-only Tensor contract follow-up

The Antigravity recommendation to add an on-device CUDA/MPS reduction is not applicable to AlbumentationsX. Runtime transforms are CPU-only, and terminal `ToTensorV2` tensors are created with `torch.from_numpy`, remain on CPU, and do not participate in autograd.

Commit `2776d8f` makes that contract explicit: the mask-survival reduction now uses the zero-copy CPU `tensor.numpy()` view, with no `.cpu()`, `.detach()`, device dispatch, GPU synchronization, or autograd handling. No non-CPU support was added. The low-severity CopyAndPaste wording was clarified separately.

Validation after this follow-up:

- targeted binding/Tensor/CopyAndPaste regressions: **25 passed**
- `uv run pytest -m "not slow"`: **12,787 passed, 42 skipped, 38 deselected, 9 xfailed, 3 xpassed**
- `uv run python -m tools.quality_gate fast`: passed, including ruff, mypy, Pyrefly, contracts, and the full pre-commit suite

The previously published [complete benchmark evidence](https://github.com/albumentations-team/AlbumentationsX/pull/384#issuecomment-5157982839) remains applicable: the NumPy direct/Compose matrix is unchanged, and its Tensor table already measures the selected zero-copy CPU `np.any(tensor.numpy())` path.

**Comment by ternaus** [comment]
## Packed `mask` invariant follow-up

The second residual risk from the latest Antigravity review is fixed in `48c936a`.

`_resync_instance_ids` now validates both supported CPU layouts before rebasing IDs:

- stacked `masks`: `len(masks) == len(bboxes)`
- packed `mask`: resolved instance-axis count equals `len(bboxes)`

A deliberately broken custom transform that drops one packed mask channel while retaining both bbox rows now raises the strict invariant error immediately. The validation was extracted into a focused helper so `_resync_instance_ids` remains below the complexity limit.

Validation:

- structural and Tensor binding regressions: **27 passed**
- `uv run pytest -m "not slow"`: **12,788 passed, 42 skipped, 38 deselected, 9 xfailed, 3 xpassed**
- `uv run python -m tools.quality_gate fast`: passed, including ruff, mypy, Pyrefly, 1,232 contract tests, and the full pre-commit suite

**Comment by ternaus** [comment]
## Final Antigravity follow-up

The device/autograd recommendation is intentionally not implemented: AlbumentationsX runtime transforms are CPU-only, terminal `ToTensorV2` outputs are CPU tensors without autograd, and this path remains the zero-copy `tensor.numpy()` reduction. No CUDA, MPS, device dispatch, `.cpu()`, or `.detach()` branches were added.

Commit `d30a906` addresses the two CPU-relevant robustness points:

- normalizes the instance axis before `index_select`
- recognizes subclasses of terminal `ToTensorV2` through their MRO, with ambiguous `N == H == W` regressions for both `transpose_mask` modes

Validation:

- targeted terminal Tensor regressions: **27 passed**
- `uv run pytest -m "not slow"`: **12,790 passed, 42 skipped, 38 deselected, 9 xfailed, 3 xpassed**
- `uv run python -m tools.quality_gate fast`: passed, including Ruff, mypy, Pyrefly, 1,232 contract tests, and full pre-commit
